[utils] Add a preliminary expression parser in eos and its test.

mreboud opened the pull request:
This is still preliminary work since Kinematics are not yet fully implemented.

dvandyk commented on eos/utils/expression-fwd.hh:
```diff
@@ -0,0 +1,33 @@
+    class ObservableNameExpression;
+    class ObservableExpression;
+
+    typedef OneOf<BinaryExpression, ConstantExpression, ObservableNameExpression, ObservableExpression> Expression;
```
Please use ``using Expression = OneOf<...>;``.

dvandyk commented on eos/utils/expression-maker.cc:
```diff
@@ -0,0 +1,72 @@
+    // copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
+    class ExpressionMaker
+    {
+        Parameters _parameters;
```
Please add an explicit ``private:`` and indent the members one level.

dvandyk commented on eos/utils/expression-maker.cc:
```diff
@@ -0,0 +1,72 @@
+                throw InternalError("Encountered ObserableExpression in ExpressionMaker::visit");
+            }
+    };
+}
```
Missing newline at the end of the file.

dvandyk commented on eos/utils/expression-parser-impl.hh:
```diff
@@ -0,0 +1,71 @@
+            | observable_name                            [ _val = _1 ]
+            ;
+
+        constant = double_ | int_;
```
I guess we will not need complex-valued numbers, since this works at the level of observables. Or am I wrong?

mreboud replied:
This is something we can easily add if needed. I would keep the parser as simple as possible until we have a complete working basis if you agree.

dvandyk replied:
I do! No complex values for now.

dvandyk commented on eos/utils/expression-parser.hh:
```diff
@@ -0,0 +1,70 @@
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> non_additive_expr;
+
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> primary_expr;
+        qi::rule<Iterator, eos::exp::ConstantExpression()       , ascii::space_type> constant;
```
Why do you specialize here to ``eos::exp::ConstantExpression``? I would leave the more general ``eos::exp::Expression``.

mreboud replied:
It simply doesn't work with `eos::exp::Expression`. I think this is needed to construct the correct class directly (with `_val=_1`).

dvandyk replied:
Understood. Keep it as is then.

dvandyk commented on eos/utils/expression-parser.hh:
```diff
@@ -0,0 +1,70 @@
+
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> primary_expr;
+        qi::rule<Iterator, eos::exp::ConstantExpression()       , ascii::space_type> constant;
+        qi::rule<Iterator, eos::exp::ObservableNameExpression() , ascii::space_type> observable_name;
```
Ditto.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+using namespace eos;
+
+
+class ExpressionEvaluator
```
This should be moved somewhere else.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+
+struct ExpressionTest
+{
+    const std::string _input;
```
The leading underscores indicate that this (and all further members) are private. But they are not. Please remove the leading underscores or make them private.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+
+    virtual void run() const
+    {
+        {
```
Please a a comment before this scope, explaining briefly what's being tested. E.g.,
```C++
// testing basic parsing of constants and binary expressions
```
or some such.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test._e.accept(printer);
```
Here the naming rang my alarm bells, since this looks like you're using a private member variable.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test._completed);
+            TEST_CHECK_EQUAL(test._e.accept_returning<double>(evaluator), 7);
```
``7.0``

mreboud replied:
In principle, this is parsed as integers, so the result is `(int) 7`. In practice, I am happy with `7.0` because `TEST_CHECK_EQUAL` doesn't test the type :)

mreboud replied:
Ah no sorry, my mistake, the evaluator evaluates to `double`.

dvandyk replied:
Hm, but your visitor can only return a double!

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,138 @@
+            Expression evaluable_test2 = test2._e.accept_returning<Expression>(maker);
+
+            // This test is to be improved with correct kinematics and default parameters
+            TEST_CHECK_RELATIVE_ERROR(evaluable_test2.accept_returning<double>(evaluator), 0.295627626, 1e-5);
```
Hm, this provide a circular dependence on the EOS libraries, but only when testing: ``libeosutils`` must be built before ``libeosbdecays.so``, but the tests of ``libeosutils`` depend on ``libeosbdecays``. Since it works, let's be happy and leave it as is.

dvandyk commented on eos/utils/expression.cc:
```diff
@@ -0,0 +1,40 @@
+
+namespace eos::exp
+{
+    double BinaryExpression::Add(double a, double b)      { return a+b; }
```
Please use lower case for all members functions (even static), except for one very particular case: named constructors.
This is not one, so please convert to lower case.

dvandyk commented on eos/utils/expression.cc:
```diff
@@ -0,0 +1,40 @@
+    double BinaryExpression::Divide(double a, double b)   { return a/b; }
+
+    BinaryExpression::func
+    BinaryExpression::Method(char op)
```
Ditto.

dvandyk commented on eos/utils/expression.cc:
```diff
@@ -0,0 +1,40 @@
+    double BinaryExpression::Add(double a, double b)      { return a+b; }
+    double BinaryExpression::Subtract(double a, double b) { return a-b; }
+    double BinaryExpression::Multiply(double a, double b) { return a*b; }
+    double BinaryExpression::Divide(double a, double b)   { return a/b; }
```
Come to think of it: ``add``, ``subtract``, ``multiply`` and ``divide`` are not expressions, they are actions. The expressions are ``sum``, ``difference``, ``product`` and ``ratio``. Care to change that?

dvandyk commented on eos/utils/expression.cc:
```diff
@@ -0,0 +1,40 @@
+            case '-': return BinaryExpression::Subtract;
+            case '*': return BinaryExpression::Multiply;
+            case '/': return BinaryExpression::Divide;
+            default:  return nullptr;
```
This should not happen, should it? I would throw an
```C++
eos::InternalError("Unknown binary operator '" + stringify(op) + "' encountered");
```

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+            char _op;
+            Expression _lhs, _rhs;
+
+            typedef double(*func)(double, double);
```
Please use ``using``:
```C++
using func = (double)(*)(double, double);
```

Why are you not using ``const double &`` for the two arguments?

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+
+            typedef double(*func)(double, double);
+
+            static double Add(double a, double b);
```
See comment above: ``sum`` vs ``add``, etc.

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+
+            BinaryExpression() {}
+
+            BinaryExpression(char op, Expression const& l, Expression const& r) :
```
EOS coding style puts ``const`` before the reference's type, i.e.: ``const Expression & l``, with all the spaces.

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+    class ConstantExpression
+    {
+        public:
+            double _value;
```
Pleasr remove the leading underscore.

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+    class BinaryExpression
+    {
+        public:
+            char _op;
```
Please remove the leading underscores.

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+        public:
+            std::string _observable_name;
+
+            ObservableNameExpression(const std::string& input = "") :
```
space between ``string`` and ``&``

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+    class ObservableExpression
+    {
+        public:
+            ObservablePtr _observable;
```
Underscore.

dvandyk commented on eos/utils/expression.hh:
```diff
@@ -0,0 +1,89 @@
+    class ObservableNameExpression
+    {
+        public:
+            std::string _observable_name;
```
Underscore.

dvandyk reviewed (changes requested):
Thank you Meril, well done! It looks very good! I have a few comments, nothing serious.

mreboud commented:
> Thank you Meril, well done! It looks very good! I have a few comments, nothing serious.

Thanks Danny for the review; I accounted for all of your comments but those concerning class specialization in the parser.

dvandyk commented:
@mreboud Can you please include my two commits for aliasing kinematic variables (``gh-dvandyk/for-meril``) in this PR?

dvandyk commented on eos/utils/kinematic.cc:
```diff
@@ -153,39 +174,86 @@ namespace eos
             result += ", " + i->first + '=' + stringify(_imp->variables_data[i->second]);
         }
 
+        // TODO: show aliases
```
Please remove this. We should not show the aliases at all.

dvandyk commented on eos/utils/kinematic.cc:
```diff
@@ -153,39 +174,86 @@ namespace eos
+
+        if (alias_exists)
+        {
+            // TODO: proper error
```
Can you please create a ``DuplicateKinematicAliasError`` in the same place where ``UnknownKinematicVariableError`` lives, and use it here? I was lazy.

dvandyk commented on eos/utils/kinematic.cc:
```diff
@@ -200,9 +268,10 @@ namespace eos
         _imp(imp),
-        _index(index)
+        _index(index),
+        _is_alias(is_alias)
```
I don't think ``_is_alias`` is required anymore now. However, I'm happy to leave it in, it might be useful later and does it does not put pressure on memory consumption.

dvandyk commented on eos/utils/kinematic_TEST.cc:
```diff
@@ -48,6 +48,30 @@ class KinematicsTest :
+                TEST_CHECK_EQUAL("s_min", kinematics["s_min"].name());
+
+                // Test clearing the alias map and access values
+                auto saved_alias = kinematics["s_min"];
```
``s_min`` is not an alias, it is **aliased*. Please add a test case that saves, in addition, ``q2_min`` and ``q2_max`` and check that they surive clearing the alias map.

dvandyk commented on eos/utils/expression-maker.cc:
```diff
@@ -0,0 +1,90 @@
+
+            Expression visit(ObservableExpression &)
+            {
+                throw InternalError("Encountered ObserableExpression in ExpressionMaker::visit");
```
Hm, is this correct? Shouldn't this simply pass through? I don't think we have any need to apply this visitor twice in a row, but it would be perfectly valid, wouldn't it?

dvandyk commented on eos/utils/expression-parser-impl.hh:
```diff
@@ -0,0 +1,111 @@
+            '['
+            >> (
+                kinematics_alias    [ phx::bind(_val, _1) ]
+              | kinematics_value    [ phx::bind(_val, _1) ]
```
I failed to indent this properly w.r.t. to the other attributes in e.g. line 75.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,142 @@
+            TEST_CHECK(test2.completed);
+            TEST_CHECK_EQUAL_STR(
+                "BinaryExpression(ObservableNameExpression(B->Dlnu::BR;l=tau, aliases=[q2_min=>q2_min_tau])\
+ / ObservableNameExpression(B->Dlnu::BR;l=mu, values=[q2_min=0]))",
```
I prefer a different multiline C string:
```
TEST_CHECK_EQUAL_STR(
    "BinaryExpression(ObservableNameExpression(B->Dlnu::BR;l=tau, aliases=[q2_min=>q2_min_tau])"
    " / "
    "ObservableNameExpression(B->Dlnu::BR;l=mu, values=[q2_min=0]))",
    out.str()
);
```

dvandyk commented on eos/utils/expression-printer.cc:
```diff
@@ -0,0 +1,84 @@
+
+            void visit(ObservableExpression &)
+            {
+                throw InternalError("Encountered ObserableExpression in ExpressionPrinter::visit");
```
Is this correct? We might want to print an expression that includes observables!

dvandyk reviewed (changes requested):
Some tiny changes, but some more important are:
  - questions regarding the treatment of ``ObservableExpression`` in ``ExpressionEvaluator`` and ``ExpressionPrinter``;
  - incorrect test case when clearing a ``Kinematics``'s aliases.

mreboud commented:
There are still a couple of steps before we can merge:
- [ ]  Fill the `_kinematics_names_array` of the class `ExpressionObservableEntry`: can be done via a visitor method.
- [ ] Implement the `uses(...)` in the class `ExpressionObservable`
- [ ] Check that object, esp. of the type `Expression`, are cloned when needed and shared when possible

dvandyk commented on eos/utils/expression-cloner.cc:
```diff
@@ -0,0 +1,90 @@
+
+            Expression visit(const ObservableExpression & e)
+            {
+                auto kinematics_values = e.kinematics_specification.values;
```
I would use ``const auto &`` here.

dvandyk commented on eos/utils/expression-cloner.cc:
```diff
@@ -0,0 +1,90 @@
+namespace eos::exp
+{
+
+    // copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
```
This comment does not reflect what the visitor does.

dvandyk commented on eos/utils/expression-maker.cc:
```diff
@@ -0,0 +1,90 @@
+                return e;
+            }
+    };
+}
```
``ExpressionCloner`` and ``ExpressionMaker`` seem to be very similar. Can we combine them into one class?

mreboud replied:
Good point

mreboud replied:
All well thought, this is maybe not a good idea. The `maker` works on `ObservableNameExpression` while the `cloner` focuses on already made `ObservableExpression`. What I can do is gather the visitors into one `expression-visitors.cc` to avoid having so many files.

dvandyk replied:
Understood!

dvandyk commented on eos/utils/expression-observable.hh:
```diff
@@ -0,0 +1,193 @@
+                _options(options),
+                _expression(expression)
+            {
+                // Uses to be implemented
```
Should we make a ``ParameterUser`` a ``ExpressionVisitor``? That would be an elegant solution to implementing parameter usage.

dvandyk reviewed (commented):
A few comments...

dvandyk commented on eos/utils/expression-parser-impl.hh:
```diff
@@ -0,0 +1,110 @@
+        observable_name =
+              '{'
+            >> as_string [ lexeme [ *~char_("}") ] ]     [ _val = _1 ]
+            >> '}'
```
Hm, we have a (small) problem here. What about
``b->smumu::Re{c9}``? We should probably:
 - use ``{{`` and ``}}`` as lead in and out for an observable name,
 - prohibit ``{{`` and ``}}`` anywhere inside a ``QualifiedName``.

dvandyk replied:
As pointed out by Meril on Discord: there are Observables that use ``}}`, hence we have to adjust. Using ``<<`` and ``>>`` now as delimiters.

mreboud replied:
I will add a commit to prohibit `<<` and `>>` in the prefix, the rest is already forbidden.

mreboud commented on eos/utils/expression-cloner.cc:
```diff
@@ -0,0 +1,88 @@
+
+                // Make observable
+                auto observable = Observable::make(
+                    e.observable->name().full(),
```
Maybe I misunderstand `full` vs `str`, but I want to be sure that all options specified in the `QualifiedName` are kept in the process.

dvandyk commented on python/_eos.cc:
```diff
@@ -298,6 +299,30 @@ BOOST_PYTHON_MODULE(_eos)
+            Represents the unit of the observables.
+
+            Seven possible entries are currently implemented in EOS:
+            Undefined, None, GeV, GeV2, InverseGeV2, InverseGeV4, InversePicoSecond.
```
Can you make this comment a list? It will be rendered as such in the documentation. You can check the documentation locally with
```
make install
make -C doc html
```
The doc will be available as ``doc/_build/html/index.html``.

dvandyk commented on python/_eos.cc:
```diff
@@ -589,6 +614,18 @@ BOOST_PYTHON_MODULE(_eos)
+            :type name: eos.QualifiedName
+            :param latex: The latex representation of the new observable.
+            :type latex: std::string
+            :param options: The set of options relevant to this new observable. These options apply to all the observables in the expression.
```
Can you emphasize the 'all' here for the documentation, i.e., ``**all**``? Generally, we can use all the markup available in reStructuredText, which is also what we're using for the parameter descriptions.

dvandyk commented on eos/observable.cc:
```diff
@@ -86,7 +88,11 @@ namespace eos
             {
-                _entries[key] = value;
+                auto inserted = _entries.insert(std::pair<QualifiedName, std::shared_ptr<const ObservableEntry>>(key, value));
+                if (!inserted.second)
```
Coding style: ``! inserted.second``. I wonder if we should rather return the boolean value, rather than ``void`` here, and let the callers of this ``ObservableEntries::insert()`` deal with potential exceptions.

dvandyk commented on eos/observable.cc:
```diff
@@ -292,8 +304,34 @@ namespace eos
 
     void
-    Observables::insert(const QualifiedName & name, const ObservableEntry * entry) const
+    Observables::insert(const QualifiedName & name, const std::string & latex, const Options & forced_options, const std::string & input_expression) const
```
Hm, the naming is a bit weird. ``input_expression`` is not an expression yet. How about ``input`` and ``expression`` in lieu of ``input_expression`` and ``parsed_expression``? Make the code a bit more readable I think.

dvandyk commented on eos/utils/expression-parser.hh:
```diff
@@ -0,0 +1,59 @@
+
+#include <eos/utils/expression.hh>
+
+namespace qi    = boost::spirit::qi;
```
Ah, this is dangerous I think, and I overlooked it so far. Can you please move these three namespace declarations inside the ``eos`` namespace?

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,200 @@
+{
+public:
+    ExpressionParserTest() :
+    TestCase("expression_parser_test")
```
Coding style: initalizations/constructor calls are indented one level (4 spaces) with respect to the class name.

dvandyk commented on eos/utils/expression-parser_TEST.cc:
```diff
@@ -0,0 +1,200 @@
+        {
+            ExpressionTest test("1 /* 2");
+
+            TEST_CHECK(!test.completed);
```
Coding style: ``! test.completed``

dvandyk commented on eos/utils/expression-visitors.cc:
```diff
@@ -0,0 +1,338 @@
+
+            Expression visit(const ObservableExpression & e)
+            {
+                return ObservableExpression(e);
```
Ahh, bug! The return value will not necessarily be bound to the same Parameters (``_parameters``). I think we need to clone the existing observable, using ``_parameters`` as the argument to clone. Maybe you were right in the first place that this case should return an error...

dvandyk commented on eos/utils/one-of.hh:
```diff
@@ -189,7 +189,11 @@ namespace eos
 
+            bool empty() const
+            {
+                return !_value;
```
Can you split this into a different commit please?

mreboud commented:
@dvandyk I rebased on master. Unless you have further comments, I suggest we merge this round and open a new issue for the future improvements, including:
 - Manage units in inserted observable entries
 - Ensure that expression observables are cacheable
 - Allow expression observable to be also added at compile time

dvandyk commented:
> @dvandyk I rebased on master. Unless you have further comments, I suggest we merge this round and open a new issue for the future improvements, including:
> 
> * Manage units in inserted observable entries
> * Ensure that expression observables are cacheable
> * Allow expression observable to be also added at compile time

Sounds good!
